File: demo/mnist_classifier.cpp
#include "IOptimizer.h"
#include "Module.h"
#include "autograd_core/expression.hpp"
#include "autograd_core/autograd_util.hpp"
#include "autograd_core/visualize.hpp"

#include "layers/Linear.h"
#include "layers/Sigmoid.h"

#include "optimizers/SGD.h"
#include "loss_functions/CrossEntropyLossWithSoftmax.h"

#include "data/IDataset.h"
#include "data/Dataloader.h"
#include "data/impl/MNIST.h"

#include "Util.h"
#include <cstdlib>
#include <memory>
#include <iostream>

class MnistClassifier : public Module {
private:
  std::unique_ptr<Linear> layers[3];

public:
  MnistClassifier() {
    this->layers[0] = std::make_unique<Linear>(28*28, 128, false);
    this->layers[1] = std::make_unique<Linear>(128, 64, false);
    this->layers[2] = std::make_unique<Linear>(64, 10, false);
  }

  virtual std::shared_ptr<autograd::Expression<Matrix>> forward(std::shared_ptr<autograd::Expression<Matrix>> ulaz) {
    for(int i=0;i < 3;i++) {
      ulaz = this->layers[i]->forward(ulaz);
      ulaz = std::make_shared<Sigmoid>(ulaz);
    }

    return ulaz;
  }
};

int main() {
  initCL_nvidia();

  std::unique_ptr<IDataset> trainset = std::make_unique<MNIST>("/home/jakov/faks/zavrsni/datasets/MNIST/raw/", true, false);
  std::unique_ptr<IDataset> testset = std::make_unique<MNIST>("/home/jakov/faks/zavrsni/datasets/MNIST/raw/", true, true);

  std::unique_ptr<Dataloader> train_loader = std::make_unique<Dataloader>(*trainset, 10);
  std::unique_ptr<Dataloader> test_loader = std::make_unique<Dataloader>(*trainset, 10);

  std::unique_ptr<MnistClassifier> classifier = std::make_unique<MnistClassifier>();
  std::shared_ptr<IOptimizer> sgd_optim = std::make_shared<SGD>(0.5f);

  int counter = 0;
  while(train_loader->hasNext()) {
    auto [input, labels] = train_loader->nextBatch();
    auto logits = classifier->forward(autograd::createVariable(input, "X"));

    auto loss = std::make_shared<CrossEntropyLossWithSoftmax>(logits, autograd::createVariable(labels, "y", false));
    classifier->backward(loss, sgd_optim);

    if(counter % 10000) {
      std::cout << "batch-logits: " << logits->getValue().toString() << std::endl;
      std::cout << "loss: " << loss->getValue().toString() << std::endl;
      exit(0);
    }

    counter++;
  }

  freeCL();
}

File: demo/mnist_classifier.py
#!/bin/python3

import torch
from torch.nn import Module
from torch import nn
from torch.utils.data import DataLoader
from torchvision import transforms
from torchvision.datasets import MNIST

class MNISTClassifier(Module):
    def __init__(self):
        super(MNISTClassifier, self).__init__()
        self.fc1 = nn.Linear(28*28, 128, False)
        self.fc2 = nn.Linear(128, 64, False)
        self.fc3 = nn.Linear(64, 10, False)

    def forward(self, input):
        out = self.fc1(input)
        out = self.fc2(out)
        out = self.fc3(out)
        return out

    def get_loss(self, X : torch.Tensor, Y_ : torch.Tensor):
        loss = nn.BCEWithLogitsLoss()
        scores = self.forward(X)

        return loss(scores, nn.functional.one_hot(Y_, 10).float())

    def get_accuracy(self, X : torch.Tensor, Y_ : torch.Tensor):
        scores = self.forward(X)
        probs = nn.functional.softmax(scores, dim=1)
        Y = probs.argmax(dim=1)
        correct = (Y == Y_).sum().float()
        total = Y_.size(dim=0)
        return float(correct) / total

if __name__ == '__main__':
    transform = transforms.Compose([
        transforms.ToTensor(),
        transforms.Lambda(lambda x: torch.flatten(x))
    ])
    trainset = MNIST(root="/home/jakov/faks/zavrsni/datasets/", train=True, transform=transform)
    testset = MNIST(root="/home/jakov/faks/zavrsni/datasets/", train=False, transform=transform)

    classifier = MNISTClassifier()
    optimizer = torch.optim.SGD(classifier.parameters(), lr=0.01)

    for epoch in range(5):
        dataloader = DataLoader(trainset, batch_size=16)
        for i, (batch_data, batch_labels) in enumerate(dataloader):
            loss = classifier.get_loss(batch_data, batch_labels)
            loss.backward()
            optimizer.step()
            optimizer.zero_grad()
            if i % 100 == 0:
                print(f"epoch: {epoch}, iteration {i}: loss {loss}")

        test_loader = DataLoader(testset, batch_size=1000, shuffle=True)
        test_sample = next(test_loader.__iter__())
        print(f"accuracy: {classifier.get_accuracy(test_sample[0], test_sample[1])}")
